File: brique-102/packages/node/src/http.ts
/**
 * HTTP client with retries, idempotency, and observability
 */
import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { MolamError } from "./errors";
import { Logger } from "./logger";
import { backoff, isRetryableStatus } from "./utils/retry";
import { ensureIdempotencyKey } from "./utils/idempotency";

export interface HttpClientOptions {
  baseUrl: string;
  apiKey: string;
  timeoutMs?: number;
  maxRetries?: number;
  userAgent?: string;
}

export class HttpClient {
  private ax: AxiosInstance;
  private opts: HttpClientOptions;

  constructor(opts: HttpClientOptions) {
    this.opts = {
      timeoutMs: 8000,
      maxRetries: 3,
      userAgent: "Molam-SDK-Node/2.0",
      ...opts
    };

    this.ax = axios.create({
      baseURL: this.opts.baseUrl,
      timeout: this.opts.timeoutMs,
      headers: {
        "Authorization": `Bearer ${this.opts.apiKey}`,
        "User-Agent": this.opts.userAgent,
        "Content-Type": "application/json"
      }
    });
  }

  private async requestWithRetry<T>(config: AxiosRequestConfig, maxRetries = 3) : Promise<T> {
    let attempt = 0;
    const idempotencyKey = ensureIdempotencyKey(config.headers as any);
    config.headers = { ...(config.headers || {}), "Idempotency-Key": idempotencyKey };

    while (true) {
      try {
        Logger.info(`HTTP ${config.method} ${config.url} attempt=${attempt+1}`);
        const res = await this.ax.request<T>(config);
        return res.data;
      } catch (err: any) {
        const status = err?.response?.status;
        const me = MolamError.fromAxios(err);
        Logger.warn(`HTTP error ${config.method} ${config.url} status=${status} attempt=${attempt+1}`, { code: me.code });

        if (attempt >= maxRetries || !isRetryableStatus(status)) {
          throw me;
        }

        const wait = backoff(attempt);
        await new Promise(r => setTimeout(r, wait));
        attempt++;
      }
    }
  }

  async get<T = any>(path: string, opts?: AxiosRequestConfig) {
    return this.requestWithRetry<T>({ method: "GET", url: path, ...opts }, this.opts.maxRetries);
  }

  async post<T = any>(path: string, body?: any, opts?: AxiosRequestConfig) {
    return this.requestWithRetry<T>({ method: "POST", url: path, data: body, ...opts }, this.opts.maxRetries);
  }

  async put<T = any>(path: string, body?: any, opts?: AxiosRequestConfig) {
    return this.requestWithRetry<T>({ method: "PUT", url: path, data: body, ...opts }, this.opts.maxRetries);
  }

  async delete<T = any>(path: string, opts?: AxiosRequestConfig) {
    return this.requestWithRetry<T>({ method: "DELETE", url: path, ...opts }, this.opts.maxRetries);
  }
}

File: brique-102/packages/go/client/http.go
package client

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/google/uuid"
)

// HttpClient handles HTTP requests with retries
type HttpClient struct {
	baseURL    string
	apiKey     string
	timeout    time.Duration
	maxRetries int
	client     *http.Client
}

// NewHttpClient creates a new HTTP client
func NewHttpClient(opts ClientOptions) *HttpClient {
	return &HttpClient{
		baseURL:    opts.BaseURL,
		apiKey:     opts.APIKey,
		timeout:    time.Duration(opts.TimeoutMS) * time.Millisecond,
		maxRetries: opts.MaxRetries,
		client:     &http.Client{Timeout: time.Duration(opts.TimeoutMS) * time.Millisecond},
	}
}

// Get makes a GET request
func (c *HttpClient) Get(path string) (map[string]interface{}, error) {
	return c.requestWithRetry("GET", path, nil)
}

// Post makes a POST request
func (c *HttpClient) Post(path string, body map[string]interface{}) (map[string]interface{}, error) {
	return c.requestWithRetry("POST", path, body)
}

// Put makes a PUT request
func (c *HttpClient) Put(path string, body map[string]interface{}) (map[string]interface{}, error) {
	return c.requestWithRetry("PUT", path, body)
}

// Delete makes a DELETE request
func (c *HttpClient) Delete(path string) (map[string]interface{}, error) {
	return c.requestWithRetry("DELETE", path, nil)
}

func (c *HttpClient) requestWithRetry(method, path string, body map[string]interface{}) (map[string]interface{}, error) {
	idempotencyKey := uuid.New().String()
	url := c.baseURL + path
	attempt := 0

	for {
		var reqBody io.Reader
		if body != nil {
			jsonData, err := json.Marshal(body)
			if err != nil {
				return nil, fmt.Errorf("failed to marshal request body: %w", err)
			}
			reqBody = bytes.NewBuffer(jsonData)
		}

		req, err := http.NewRequest(method, url, reqBody)
		if err != nil {
			return nil, fmt.Errorf("failed to create request: %w", err)
		}

		req.Header.Set("Authorization", "Bearer "+c.apiKey)
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("User-Agent", "Molam-SDK-Go/2.0")
		req.Header.Set("Idempotency-Key", idempotencyKey)

		resp, err := c.client.Do(req)
		if err != nil {
			if attempt >= c.maxRetries {
				return nil, fmt.Errorf("request failed after %d retries: %w", c.maxRetries, err)
			}
			time.Sleep(backoff(attempt))
			attempt++
			continue
		}

		defer resp.Body.Close()

		respBody, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, fmt.Errorf("failed to read response body: %w", err)
		}

		if resp.StatusCode >= 400 {
			if attempt < c.maxRetries && isRetryableStatus(resp.StatusCode) {
				time.Sleep(backoff(attempt))
				attempt++
				continue
			}

			var errResp map[string]interface{}
			json.Unmarshal(respBody, &errResp)
			return nil, fmt.Errorf("API error: %d - %v", resp.StatusCode, errResp)
		}

		var result map[string]interface{}
		if len(respBody) > 0 {
			if err := json.Unmarshal(respBody, &result); err != nil {
				return nil, fmt.Errorf("failed to unmarshal response: %w", err)
			}
		}

		return result, nil
	}
}

func isRetryableStatus(status int) bool {
	if status >= 500 {
		return true
	}
	return status == 408 || status == 429 || status == 425
}

func backoff(attempt int) time.Duration {
	sequence := []int{200, 500, 1000, 2000, 5000}
	idx := attempt
	if idx >= len(sequence) {
		idx = len(sequence) - 1
	}
	return time.Duration(sequence[idx]) * time.Millisecond
}
